sdk/v1/composable.go

```diff
@@ -0,0 +1,409 @@
+			return nil, compErr
+		}
+	} else { //labelsOK
+		strLabels := make(map[string]string)
+		for key, value := range intLabels {
+			strValue := fmt.Sprintf("%v", value)
+			strLabels[key] = strValue
+		}
+		unstrList := unstructured.UnstructuredList{}
+		unstrList.SetGroupVersionKind(groupVersionKind)
+		err = r.List(context.TODO(), &unstrList, client.InNamespace(ns), client.MatchingLabels(strLabels))
+		if err != nil {
+			logf.Info("list object returned ", "err", err, "namespace", ns, "labels", strLabels, "groupVersionKind", groupVersionKind)
+			compErr = &ComposableError{err, true, true}
+			cache.objects[key] = toumbstone{err: *compErr}
+			return nil, compErr
+		}
+		itms := len(unstrList.Items)
+		if itms == 1 {
+			unstrObj = unstrList.Items[0]
+			cache.objects[key] = unstrObj
+		} else {
+			err = fmt.Errorf("list object returned %d items ", itms)
+			logf.Error(err, "wrong # of items", "items", itms, "namespace", Namespace, "labels", strLabels, "groupVersionKind", groupVersionKind)
+			compErr = &ComposableError{err, true, true}
+			cache.objects[key] = toumbstone{err: *compErr}
+			return nil, compErr
+		}
+	}
+	return &unstrObj, nil
+}
+
+func resolveValue2(val map[string]interface{}, unstrObj unstructured.Unstructured, path string) (interface{}, *ComposableError) {
+	j := jsonpath.New("compose")
+	// add ".Object" to the path
+	path = path[:1] + objectPrefix + path[1:]
+	err := j.Parse(path)
+	if err != nil {
+		logf.Error(err, "jsonpath.Parse", "path", path)
+		return nil, &ComposableError{err, false, false}
+	}
+	j.AllowMissingKeys(false)
+
+	fullResults, err := j.FindResults(unstrObj)
+	if err != nil {
+		logf.Error(err, "FindResults", "obj", unstrObj, "path", path)
+		if strings.Contains(err.Error(), "is not found") {
+			return errorToDefaultValue(val, ComposableError{err, true, true})
+		}
+		return nil, &ComposableError{err, false, false}
+	}
+	iface, ok := template.PrintableValue(fullResults[0][0])
+	if !ok {
+		err = fmt.Errorf("can't find printable value %v ", fullResults[0][0])
+		logf.Error(err, "template.PrintableValue", "obj", unstrObj, "path", path)
+		return nil, &ComposableError{err, false, false}
+	}
+
+	var retVal interface{}
+	if transformers, ok := val[Transformers].([]interface{}); ok && len(transformers) > 0 {
+		transformNames := make([]string, 0, len(transformers))
+		for _, v := range transformers {
+			if name, ok := v.(string); ok {
+				transformNames = append(transformNames, name)
+			}
+		}
+		retVal, err = CompoundTransformerNames(iface, transformNames...)
+	} else {
+		retVal = iface
+	}
+	return retVal, nil
+}
+
+func errorToDefaultValue(val map[string]interface{}, err ComposableError) (interface{}, *ComposableError) {
+	if defaultValue, ok := val[defaultValue]; ok {
+		return defaultValue, nil
+	}
+	return nil, &err
+}
+
+func getDiscoveryClient(cfg *rest.Config) discovery.CachedDiscoveryInterface {
+	return memory.NewMemCacheClient(discovery.NewDiscoveryClientForConfigOrDie(cfg))
```

Comment:
You recreate the CacheDiscoveryClient per each `lookupAPIResource` call which is done per each input Object resolution.
If you don't want to use the REST API cache, you can use just `DiscoveryClient` 
<
import "k8s.io/client-go/discovery"
....
DiscoveryClient := discovery.NewDiscoveryClientForConfigOrDie(mgr.GetConfig())
>
However, you might faced with a race condition between different `lookupAPIResource` during the same reconciliation call. If you want, you can implement the caching by yourself, as I did before K8s 1.14
https://github.com/IBM/composable/blob/95378c481b0368b6f05355eae7a654c4abeda8b4/pkg/controller/composable/composable_controller.go#L76

Reply:
DiscoveryClient is now created per Resolve and not lookupAPIResource